oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java

```diff
@@ -0,0 +1,199 @@
+ * ComputeEngineCredentials, ImpersonatedCredentials.
+ *
+ * <p>For more information see <br>
+ * Usage:<br>
+ *
+ * <pre>
+ * String credPath = "/path/to/svc_account.json";
+ * String targetAudience = "https://example.com";
+ *
+ * // For Application Default Credentials (as ServiceAccountCredentials)
+ * // export GOOGLE_APPLICATION_CREDENTIALS=/path/to/svc.json
+ * GoogleCredentials adcCreds = GoogleCredentials.getApplicationDefault();
+ * if (!adcCreds instanceof IdTokenProvider) {
+ *   // handle error message
+ * }
+ * 
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.newBuilder()
+ *     .setIdTokenProvider(adcCreds)
+ *     .setTargetAudience(targetAudience).build();
+ * 
+ * // for ServiceAccountCredentials
+ * ServiceAccountCredentials saCreds = ServiceAccountCredentials.fromStream(new FileInputStream(credPath));
+ * saCreds = (ServiceAccountCredentials) saCreds.createScoped(Arrays.asList("https://www.googleapis.com/auth/iam"));
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.newBuilder()
+ *     .setIdTokenProvider(saCreds)
+ *     .setTargetAudience(targetAudience).build();
+ * 
+ * // for ComputeEngineCredentials
+ * ComputeEngineCredentials caCreds = ComputeEngineCredentials.create();
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.newBuilder()
+ *     .setIdTokenProvider(caCreds)
+ *     .setTargetAudience(targetAudience)
+ *     .setOptions(Arrays.asList(ComputeEngineCredentials.ID_TOKEN_FORMAT_FULL))
+ *     .build();
+ * 
+ * // for ImpersonatedCredentials
+ * ImpersonatedCredentials imCreds = ImpersonatedCredentials.create(saCreds,
+ *     "impersonated-account@project.iam.gserviceaccount.com", null,
+ *     Arrays.asList("https://www.googleapis.com/auth/cloud-platform"), 300);
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.newBuilder()
+ *     .setIdTokenProvider(imCreds)
+ *     .setTargetAudience(targetAudience)
+ *     .setOptions(Arrays.asList(ImpersonatedCredentials.INCLUDE_EMAIL))
+ *     .build();
+ *
+ * // Use the IdTokenCredential in an authorized transport
+ * GenericUrl genericUrl = new GenericUrl("https://example.com");
+ * HttpCredentialsAdapter adapter = new HttpCredentialsAdapter(tokenCredential);
+ * HttpTransport transport = new NetHttpTransport();
+ * HttpRequest request = transport.createRequestFactory(adapter).buildGetRequest(genericUrl);
+ * HttpResponse response = request.execute();
+ *
+ * // Print the token, expiration and the audience
+ * System.out.println(tokenCredential.getIdToken().getTokenValue());
+ * System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getAudienceAsList());
+ * System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getExpirationTimeSeconds());
+ * </pre>
+ */
+@Beta
+public class IdTokenCredentials extends OAuth2Credentials {
+
+  private static final long serialVersionUID = -2133257318957588431L;
+
+  private IdTokenProvider idTokenProvider;
```

Comment:
All of these are nullable. Is that what you want?

Reply:
so do the checks here?
```java
  private IdTokenCredentials(Builder builder) {
    this.idTokenProvider = Preconditions.checkNotNull(builder.getIdTokenProvider());
    this.targetAudience = Preconditions.checkNotNull(builder.getTargetAudience());
    this.options = Preconditions.checkNotNull(builder.getOptions());
  }
```

Reply:
That works.

Reply:
ok, the options can be nullable so removed the condition checks on that